File: array/problem80/problem80.py
from typing import List


class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        # k points to the next write position
        k = 0
        for x in nums:
            # keep x if we have written fewer than 2 copies of x
            # or x differs from the element written two positions back
            if k < 2 or x != nums[k - 2]:
                nums[k] = x
                k += 1
        return k

File: array/problem80/problem80.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        int k = 0;
        for (int x : nums) {
            // Keep x if we have fewer than 2 kept elements of this value,
            // or if x differs from the element kept two positions ago.
            if (k < 2 || x != nums[k - 2]) {
                nums[k] = x;
                ++k;
            }
        }
        return k;
    }
};
